compiler/rustc_codegen_llvm/src/intrinsic_util.rs

```diff
@@ -0,0 +1,180 @@
+                    return;
+                }
+
+                if adt_def.is_enum() {
+                    // This terminates the current basic block with a switch
+                    // statement, so the builder must be updated to use the
+                    // successor.
+                    *builder = codegen_collectable_calls_for_enum(builder, place, substs, *adt_def);
+                    return;
+                }
+
+                // Iterate over the adt fields to ensure that if their values
+                // contain Collectable implementations, they are also called.
+                for i in 0..place.layout.fields.count() {
+                    let field = place.project_field(builder, i);
+                    codegen_collectable_calls_for_value(builder, field, substs);
+                }
+            },
+        _ => todo!(),
+    }
+}
+
+// If an enum variant contains type(s) which implement `Collectable`, the
+// `set_collectable` method must only be called if that variant is instantiated.
+// In general, this can only be known at runtime, so codegenning collectable
+// calls for an enum requires inserting a switch statement which branches on the
+// discriminant.
+//
+// The basic algorithm works as follows:
+//
+//    1. A successor basic block is created, where control flow converges after
+//       the switch statement. This successor is the block associated with the
+//       `builder` for all future codegen'd statements in the same
+//       `make_collectable` intrinsic.
+//
+//    2. A basic block is created for each variant, these dominate the successor
+//       block.
+//
+//    3. The fields of each variant are iterated over and have their types
+//       recursed into in order to look for more `Collectable` impls. Once
+//       found, calls to `set_collectable` methods are codegen'd to the
+//       respective variant block.
+//
+//    4. The builder is "patched up" to use the successor block.
+//
+// This is recursive, and can contain nested switch statements of arbitrary
+// depth where enum variants contain other enums. There's a potential here for
+// an explosion in the complexity of the CFG because of all the layers of nested
+// switches and blocks, but I think this goes away when the LLVM IR is lowered
+// into the target backend.
+fn codegen_collectable_calls_for_enum<'a, 'll, 'tcx>(
+    builder: &mut Builder<'a, 'll, 'tcx>,
+    place: PlaceRef<'tcx, &'ll Value>,
+    substs: SubstsRef<'tcx>,
+    adt_def: &'tcx AdtDef,
+) -> Builder<'a, 'll, 'tcx> {
+    // Codegen loading the enum discriminant.
+    let discr_ty = adt_def.repr.discr_type().to_ty(builder.tcx);
+    let discr = place.codegen_get_discr(builder, discr_ty);
+    let discr_rv = OperandRef {
+        val: OperandValue::Immediate(discr),
+        layout: builder.cx().layout_of(discr_ty),
+    };
+
+    // Build the successor block.
+    let name = format!("successor_{:?}", adt_def.did.index);
```

Comment:
Do these names (IDs?) have to be locally or globally unique?

Reply:
Gah, globally I think. I'll fix this.

Reply:
I think we just need the crate hash in here then?

Reply:
It's enough to just serialize the `DefId` because they are globally unique. The hashing API is weird and I'm not sure I fully understand it. Fix here: 740f974eb077097c682f717a2bd9929d6c9ba0b8